activity_calendar/models.py

```diff
@@ -22,45 +30,232 @@ class Meta:
+    # Maximum number of participants
+    def get_max_num_participants(self, recurrence_id=None):
+        max_participants = self.max_participants
+
+        # At least one slot can (in theory) be created
+        if self.slot_creation != "CREATION_NONE" and self.max_slots != 0:
+            # New slots can actually be made (take into account the current limit)
+            if self.max_slots < 0 or self.max_slots - self.get_num_slots(recurrence_id) > 0:
+                # Only limited by this activity's participants
+                return max_participants
+
+        # Otherwise we have to deal with the limitations of the already existing slots
+        cnt = 0
+        for slot in self.get_slots(recurrence_id):
+            # At least one slot allows for infinite participants
+            if slot.max_participants < 0:
+                # But may still be limited by the activity's maximum amount of participants
+                return max_participants
+            cnt += slot.max_participants 
+        
+        if max_participants < 0:
+            # Infinite activity participants means we're limited by the existing slots
+            return cnt
+        # Otherwise it's the smallest of the two
+        return min(max_participants, cnt)
+
+    # slots already created
+    def get_slots(self, recurrence_id=None):
+        if not self.is_recurring:
+            return ActivitySlot.objects.filter(parent_activity__id=self.id)
+
+        if recurrence_id is None:
+            raise TypeError("recurrence_id cannot be None if the activity is recurring")
+
+        return ActivitySlot.objects.filter(parent_activity__id=self.id,
+                recurrence_id=recurrence_id)
+    
+    # Number of slots
+    def get_num_slots(self, recurrence_id=None):
+        return self.get_slots(recurrence_id).count()
+
+    # Maximum number of slots that can be created
+    def get_max_num_slots(self, recurrence_id=None):
+        # Users can create at least one slot
+        if self.slot_creation != "CREATION_NONE":
+            return self.max_slots
+
+        # Slots cannot be created outside the admin panel; we have to work with
+        # the slots that already exist
+        return self.get_num_slots(self, recurrence_id)
+    
+    def can_user_create_slot(self, user, recurrence_id=None, num_slots=None, num_user_registrations=None,
+            num_total_participants=None, num_max_participants=None):
+        if num_user_registrations is None:
+            num_user_registrations = self.get_num_user_subscriptions(user, recurrence_id=recurrence_id)
+        
+        if num_total_participants is None:
+            self.get_slots(recurrence_id=recurrence_id).aggregate(Count('participants'))['participants__count']
+        if num_max_participants is None:
+            self.get_max_num_participants(recurrence_id=recurrence_id)
+
+        # Can the user (in theory) join another slot?
+        user_can_join_another_slot = (self.max_slots_join_per_participant < 0 or \
```

Comment:
If you are using -1 as a magic value, please do `== -1` and not `< 0` because this isn't really clear